backend/tilavarauspalvelu/api/graphql/types/reservation/serializers/create_serializer.py

```diff
@@ -118,16 +118,50 @@ def prefill_reservation_from_profile(self, data: ReservationCreateData) -> None:
         if prefill_info is not None:
             data.update({key: value for key, value in prefill_info.items() if value is not None})
 
-    @transaction.atomic()
     def create(self, validated_data: ReservationCreateData) -> Reservation:
-        # Must be able to link reservation unit to the reservation
-        reservation_unit: ReservationUnit = validated_data.pop("reservation_unit")
-        reservation: Reservation = super().create(validated_data)
-        reservation.reservation_units.set([reservation_unit])
-
-        # Pindora request must succeed, or the reservation is not created.
+        with transaction.atomic():
+            # Must be able to link reservation unit to the reservation
+            reservation_unit: ReservationUnit = validated_data.pop("reservation_unit")
+            reservation: Reservation = super().create(validated_data)
+            reservation.reservation_units.set([reservation_unit])
+
+        begin = reservation.begin.astimezone(DEFAULT_TIMEZONE)
+        end = reservation.end.astimezone(DEFAULT_TIMEZONE)
+
+        # After creating the reservation, check again if there are any overlapping reservations.
+        # This can fail if two reservations are created for reservation units in the same
+        # space-resource hierarchy at almost the same time, meaning when we check for overlapping
+        # reservations during validation, neither of the reservations are yet created.
+        try:
+            reservation_unit.validators.validate_no_overlapping_reservations(
+                begin=begin, end=end, ignore_ids=[reservation.id]
+            )
+        except ValidationError:
+            overlapping_reservations = reservation.actions.overlapping_reservations()
+            for overlapping_reservation in overlapping_reservations:
+                #
+                # Remove this reservation if it was created later than any of the overlapping reservations.
+                if reservation.created_at > overlapping_reservation.created_at:
```

Comment:
To be pedantic, I suppose the comparison should strictly speaking be `≥` instead of `>` (is it at least theoretically possible that the timestamps are assigned the same value at the precision supported?) -- but then how to resolve which was first -- could the comparison then be by the `id` instead since the lower should be the first?

In any case I guess the "winner" hinges on which request gets to open the DB transaction first -- and that might not actually be the first request received for processing -- so the outcome is in that sense somewhat arbitrary anyway (outside this test setup which is deterministic).

Reply:
Changed this to remove regardless, according to your later comment